### ISSUE TITLE
twisted.web.server.Request.processingFailed can return a str, not bytes, on Py3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7996]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7996|
|Type|defect|
|Created|2015-08-10 06:22:32Z|
|Branch|https://github.com/twisted/twisted/tree/processingfailed-py3-7996|

If ```site.displayTracebacks``` is True on Py3, and the processing failed, then it will try and write str to the request, not bytes.

<details><summary>Searchable metadata</summary>

```
trac-id__7996 7996
type__defect defect
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_processingfailed_py3_7996 branches/processingfailed-py3-7996
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1439187752936033 1439187752936033
changetime__1439613800005341 1439613800005341
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [45427]) Branching to processingfailed-py3-7996.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

Builders are looking green enough. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @hawkowl|
|-|-|

For test test_processingFailed I think that the docstring should state that the failure is written as utf-8 in HTML, only when displayTracebacks is enabled :)

Maybe the test should be renamed test_processingFailed_displayTracebacks

Also, I see that your changes modified the content-type header and the content encoding... but there is no test for that. Maybe test_processingFailed should be extended to check for contet-type / content-lenght headers and use an exception with non-ascii test.

Please check my command and resubmit.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

Okay, I've pared some things back and added some tests. Builders look green.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes look good.

why are you using 'self.assertEqual(1, len(self.flushLoggedErrors()))' in tests?
I don't see anything in test's docstring to suggest why we need it.

Maybe it can be extracted into a separate helper method which a suggestive name and docstring.

Please check my feedback and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

Hi Adi, thanks for the review.

This is because otherwise the test fails because there was an uncaptured exception, because we make one in the test which didn't get 'raised'.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

Please add a comment or some in code hints so that when someone reads this code and has a WTF moment, will not have to dig the commit history and check all the review comments.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45474]) Merge processingfailed-py3-7996: Make twisted.web.server.Request.processingFailed work on Python 3

Author: hawkowl
Reviewer: adiroiban
Fixes: [#7996](https://github.com/twisted/twisted/issues/7996)
